packages/graphql-eslint-rules/src/missing-apollo-key-fields.ts

```diff
@@ -0,0 +1,263 @@
+// Based on https://github.com/dotansimha/graphql-eslint/blob/%40graphql-eslint/eslint-plugin%401.1.4/packages/plugin/src/rules/require-id-when-available.ts
+
+import {
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLOutputType,
+  GraphQLNamedType,
+  isNonNullType,
+  isListType,
+} from "graphql";
+
+const REQUIRE_KEY_FIELDS_WHEN_AVAILABLE = "missing-apollo-key-fields";
+const DEFAULT_KEY_FIELD_NAME = "id";
+
+export function getBaseType(type: GraphQLOutputType): GraphQLNamedType {
+  if (isNonNullType(type) || isListType(type)) {
+    return getBaseType(type.ofType);
+  }
+
+  return type;
+}
+
+function checkRequiredParameters(ruleName: string, context: any) {
+  if (!context.parserServices) {
+    throw new Error(
+      `Rule '${ruleName}' requires 'parserOptions.operations' to be set and loaded. See http://bit.ly/graphql-eslint-operations for more info`
+    );
+  }
+
+  if (!context.parserServices.siblingOperations.available) {
+    throw new Error(
+      `Rule '${ruleName}' requires 'parserOptions.operations' to be set and loaded. See http://bit.ly/graphql-eslint-operations for more info`
+    );
+  }
+
+  if (!context.parserServices.hasTypeInfo) {
+    throw new Error(
+      `Rule '${ruleName}' requires 'parserOptions.schema' to be set and schema to be loaded. See http://bit.ly/graphql-eslint-schema for more info`
+    );
+  }
+
+  if (!context.options[0]?.typePolicies) {
+    throw new Error(
+      `Rule '${ruleName}' requires option 'typePolicies' to be set.`
+    );
+  }
+}
+
+function keyFieldsForType(
+  type: GraphQLObjectType | GraphQLInterfaceType,
+  typePolicies: any
+) {
+  const keyFields: string[] = [];
+  const typePolicy = typePolicies[type.name];
+  if (typePolicy && typePolicy.keyFields) {
+    if (Array.isArray(typePolicy.keyFields)) {
+      typePolicy.keyFields.forEach((keyField: any) => {
+        if (typeof keyField === "string") {
+          keyFields.push(keyField);
+        } else {
+          throw new Error("Expected keyFields to be array of strings");
+        }
+      });
+    } else {
+      throw new Error("Expected keyFields to be array of strings");
+    }
+  } else if (type.getFields().id !== undefined) {
+    keyFields.push(DEFAULT_KEY_FIELD_NAME);
+  }
+  return keyFields;
+}
+
+function getKeyFieldsObjectForCheck(keyFields: string[]) {
+  return keyFields.reduce((acc, id) => {
+    acc[id] = false;
+    return acc;
+  }, {} as Record<string, boolean>);
+}
+
+function getUnusedKeyFields(idsObjectForCheck: Record<string, boolean>) {
+  return Object.entries(idsObjectForCheck).reduce((acc, [key, value]) => {
+    if (value) {
+      return acc;
+    }
+    acc.push(key);
+    return acc;
+  }, [] as string[]);
+}
+
+export default {
+  meta: {
+    type: "suggestion",
```

Comment:
What does this signify?

Reply:
type of the rule ... "problem" is probably better as long as we want to enforce it.

Reply:
Yes, ours is not a suggestion